_posts/2021-01-09-Julia-1.6-what-has-changed-since-1.0.md

```diff
@@ -0,0 +1,635 @@
+
+With the new tiered resolved, `pkg> add` will try a number of increasingly relaxed strategies to try and install the new package while making minimal changes to the versions of the existing packages that are already installed.
+You can see the full description of the [`--preserve` options in the Pkg docs](https://julialang.github.io/Pkg.jl/v1/repl/#repl-add).
+
+ - First it will try and avoid all changes.
+ - Then it will try and avoid changing the version of direct dependencies, but allow changes to the version of indirect dependencies.
+ - Then it will allow changes to direct dependencies but will avoid changing the major (or minor pre-1.0) version, even if `Project.toml`’s `[compat]` says it is allowed to.
+ - Then it will allow changes to everything (as long as permitted by `Project.toml`’s `[compat]` section)
+ - Finally it will give up and error if it still can’t find a compatible set of package that let it add the new one.
+
+For a package with a fully filled in `[compat]` section in the  `Project.toml`, then it is always fine to allow changes to everything -- which is what this will do if it has to.
+If one hasn't setup the `[compat]` section then this can be nice to minimize changes.
+(But also: please setup your `[compat]` bounds, for your own sake.)
+
+The net result of this is far more intuitive behavior for `pkg> add`.
+If there is a compatible set of package versions then they will be found, regardless of whether new packages are added all at once or one at a time.
+
+### Precompilation
+
+Precompilation has been enhanced a lot.
+That is not the compilation that runs the first time a function is used in a session, but rather the stuff that runs the first time a package is loaded in an environment.
+
+For a start, the precompilation cache no longer goes stale every time you swap environments.
+This was a massive pain in Julia 1.0, especially if you worked on more than one thing at a time.
+This was fixed in 1.3 to have a separate cache for each environment.
+It's easy to forget this one, but it is actually one of the biggest usability enhancements since 1.0.
+
+More dramatic, and less easy to overlook, is the parallelism of precompilation added in 1.6.
+Precompiling a package requires precompiling all its dependencies first.
+This is now done in parallel, and is automatically triggered when you complete Pkg operations, in contrast to happening serially the first time a package is loaded.
+No more waiting 5 minutes the first time you load a package with a lot of dependencies.
+Further, the spiffy animated output shows you what is precompiling at any given time, as well as a progress bar, which makes it feel (even) faster.
+
+<script id="asciicast-a26toW3opklSEtaRAeXzpcW8W" src="https://asciinema.org/a/a26toW3opklSEtaRAeXzpcW8W.js" async></script>
```

Comment:
Maybe [this YouTube video](https://www.youtube.com/watch?v=9SHuhIGfVDg) is more impressive?